glommio/src/sync/semaphore.rs

```diff
@@ -7,49 +7,115 @@ use std::cell::RefCell;
 use std::future::Future;
 use std::io::{Error, ErrorKind, Result};
 use std::pin::Pin;
-use std::rc::Rc;
 use std::task::{Context, Poll, Waker};
 
-use intrusive_collections::intrusive_adapter;
+use intrusive_collections::linked_list::LinkOps;
+use intrusive_collections::{container_of, offset_of, Adapter, PointerOps};
 use intrusive_collections::{LinkedList, LinkedListLink};
+use std::marker::PhantomPinned;
+use std::ptr::NonNull;
 
 #[derive(Debug)]
-struct Waiter {
-    node: Rc<WaiterNode>,
-    sem_state: Rc<RefCell<SemaphoreState>>,
+struct Waiter<'a> {
+    node: WaiterNode,
+    semaphore: &'a Semaphore,
 }
 
 #[derive(Debug)]
 struct WaiterNode {
     link: LinkedListLink,
-    state: RefCell<WaiterState>,
+    units: u64,
+    waker: RefCell<Option<Waker>>,
+
+    //waiter node can not be Unpin so its pointer could be used inside of intrusive
+    //collection, it also can not outlive the container which is guaranteed by the
+    //Waiter lifetime bound to the Semaphore which is container of all Waiters.
+    _p: PhantomPinned,
 }
 
-#[derive(Debug)]
-struct WaiterState {
-    units: u64,
-    waker: Option<Waker>,
+struct WaiterPointerOps;
+
+unsafe impl PointerOps for WaiterPointerOps {
+    type Value = WaiterNode;
+    type Pointer = NonNull<WaiterNode>;
+
+    unsafe fn from_raw(&self, value: *const Self::Value) -> Self::Pointer {
+        NonNull::new_unchecked(value as *mut Self::Value)
+    }
+
+    fn into_raw(&self, ptr: Self::Pointer) -> *const Self::Value {
+        ptr.as_ptr() as *const Self::Value
+    }
 }
 
-intrusive_adapter!(WaiterAdapter = Rc<WaiterNode> : WaiterNode {link : LinkedListLink});
+struct WaiterAdapter {
+    pointers_ops: WaiterPointerOps,
+    link_ops: LinkOps,
+}
+
+impl WaiterAdapter {
+    fn new() -> Self {
+        WaiterAdapter {
+            pointers_ops: WaiterPointerOps,
+            link_ops: LinkOps,
+        }
+    }
+}
+
+unsafe impl Adapter for WaiterAdapter {
+    type LinkOps = LinkOps;
+    type PointerOps = WaiterPointerOps;
+
+    unsafe fn get_value(
+        &self,
+        link: <Self::LinkOps as intrusive_collections::LinkOps>::LinkPtr,
+    ) -> *const <Self::PointerOps as PointerOps>::Value {
+        container_of!(link.as_ptr(), WaiterNode, link)
+    }
 
-impl Waiter {
-    fn new(units: u64, sem_state: Rc<RefCell<SemaphoreState>>) -> Waiter {
+    unsafe fn get_link(
+        &self,
+        value: *const <Self::PointerOps as PointerOps>::Value,
+    ) -> <Self::LinkOps as intrusive_collections::LinkOps>::LinkPtr {
+        let ptr = (value as *const u8).add(offset_of!(WaiterNode, link));
+        core::ptr::NonNull::new_unchecked(ptr as *mut _)
```

Comment:
that doesn't seem effective:
if `Value` is null, then the final ptr will be Null + offset which will pass the non_null check but still be invalid.

Why don't we have to check the original ptr for non-null ?
While we're at it, doesn't using `new_unchecked` defeats the whole purpose of using `NonNull` ?

Reply:
Hi @glommer , those pointers can not be null because this trait implementation is used by intrusive_collections to convert pointers from one form to another (pointer to the link field to the pointer to the value field and vice versa). So those methods always receive valid pointers.

Reply:
The container itself guarantees validness of the pointer but I suppose additional check will not harm.